lib/widgets/editor/editor_toolbar.dart

```diff
@@ -0,0 +1,462 @@
+import '../../util/text_lines_iterator.dart';
+import 'editor_picking_dialog.dart';
+import 'editor_toolbar_store.dart';
+
+class Reformat {
+  final String text;
+  final int selectionBeginningShift;
+  final int selectionEndingShift;
+  Reformat({
+    required this.text,
+    this.selectionBeginningShift = 0,
+    this.selectionEndingShift = 0,
+  });
+}
+
+extension on TextEditingController {
+  String get selectionText =>
+      text.substring(selection.baseOffset, selection.extentOffset);
+  String get beforeSelectionText => text.substring(0, selection.baseOffset);
+  String get afterSelectionText => text.substring(selection.extentOffset);
+
+  /// surroungs selection with given strings. If nothing is selected, placeholder is used in the middle
+  void surround({
+    required String before,
+    required String placeholder,
+
+    /// after = before if null
+    String? after,
+  }) {
+    after ??= before;
+    final beg = text.substring(0, selection.baseOffset);
+    final mid = () {
+      final m = text.substring(selection.baseOffset, selection.extentOffset);
+      if (m.isEmpty) return placeholder;
+      return m;
+    }();
+    final end = text.substring(selection.extentOffset);
+
+    value = value.copyWith(
+        text: '$beg$before$mid$after$end',
+        selection: selection.copyWith(
+          baseOffset: selection.baseOffset + before.length,
+          extentOffset: selection.baseOffset + before.length + mid.length,
+        ));
+  }
+
+  String get firstSelectedLine {
+    if (text.isEmpty) return '';
+    return text.substring(text.getBeginningOfTheLine(selection.start - 1),
+        text.getEndOfTheLine(selection.end));
+  }
+
+  void insertAtBeginningOfFirstSelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this)..moveNext();
+    lines.current = s + lines.current;
+    value = value.copyWith(
+      text: lines.text,
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset + s.length,
+        extentOffset: selection.extentOffset + s.length,
+      ),
+    );
+  }
+
+  void removeAtBeginningOfEverySelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this);
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (lines.current.startsWith(RegExp.escape(s))) {
+          lines.current = lines.current.substring(s.length);
+          linesCount++;
+        }
+      }
+    }
+
+    value = value.copyWith(
+      text: lines.text,
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset - s.length,
+        extentOffset: selection.extentOffset - s.length * linesCount,
+      ),
+    );
+  }
+
+  void insertAtBeginningOfEverySelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this);
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (!lines.current.startsWith(RegExp.escape(s))) {
```

Comment:
Why RegExp.escape here?

Reply:
startswith parses the string as regex

Reply:
Are you sure? https://api.flutter.dev/flutter/dart-core/String/startsWith.html
It accepts a Pattern, where a normal string is a literal pattern

Reply:
well now i see that i thought wrong for a very long time 😅